### PR TITLE
nixos/blocky: init

### PR DESCRIPTION
**Pull request by ratsclub**
###### Motivation for this change
Added the module needed to use blocky.

See: https://github.com/NixOS/nixpkgs/pull/156655

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/blocky.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+let
+  cfg = config.services.blocky;
+
+  format = pkgs.formats.yaml { };
+  configFile = format.generate "config.yaml" cfg.settings;
+in
+{
+  options.services.blocky = {
+    enable = mkEnableOption ''
+      Fast and lightweight DNS proxy as ad-blocker for local network with many features.
```
Comment (aanderse) [inline nixos/modules/services/networking/blocky.nix:14]: `mkEnableOption` includes a trailing period so you can omit that.

**Comment by aanderse** [comment]
Any chance for a NixOS test? We probably have a few dns ones already, easily adaptable.

**Comment by ratsclub** [comment]
Do you think this small test is enough?

File: nixos/tests/blocky.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            default = [ "8.8.8.8" "1.1.1.1" ];
+          };
+          port = 53;
+          httpPort = 4000;
+          logLevel = "info";
+        };
+      };
+    };
+
+    mixedConf = { ... }: {
```
Comment (aanderse) [inline nixos/tests/blocky.nix:20]: I don't understand what this is for. The description mentions merging configuration... there is no merge here.

**Review by aanderse — COMMENTED** [review]
> Do you think this small test is enough?

Maybe add something like [this](https://github.com/0xERR0R/blocky/blob/64c198529136b07dfa9146c4a545e9481f2e06e7/docs/config.yml#L18-L23):
```yml
customDNS:
  customTTL: 1h
  mapping:
    printer.lan: 192.168.178.3,2001:0db8:85a3:08d3:1319:8a2e:0370:7344
```

Then you could run `dig @127.0.0.1 printer.lan`, etc...

File: nixos/tests/blocky.nix (hunk 2 above)
Comment (ratsclub) [reply nixos/tests/blocky.nix:20]: > ```yaml
> customDNS:
>   customTTL: 1h
>   mapping:
>     printer.lan: 192.168.178.3,2001:0db8:85a3:08d3:1319:8a2e:0370:7344
> ```
> 
> Then you could run `dig @127.0.0.1 printer.lan`, etc...

I'll try to do something like this. This is my first time doing network related packaging, thanks for pointing it out :sweat_smile:

**Comment by aanderse** [comment]
FYI - I have been running this module for simple DNS needs locally on my network for a bit... works great :+1:

**Comment by ratsclub** [comment]
@aanderse what do you think about it now?

File: nixos/tests/blocky.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import ./make-test-python.nix {
+  name = "blocky";
+
+  nodes = {
+    minimalConf = { ... }: {
```
Comment (aanderse) [inline nixos/tests/blocky.nix:5]: I think a single machine is fine. You can drop this one.

File: nixos/tests/blocky.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    with subtest("Minimal config test"):
+        minimalConf.wait_for_unit("blocky.service")
+        minimalConf.wait_for_open_port(53)
+        minimalConf.wait_for_open_port(4000)
+
+    with subtest("Bigger config test"):
+        biggerConf.wait_for_unit("blocky.service")
+        biggerConf.wait_for_open_port(53)
+        biggerConf.wait_for_open_port(5000)
+        biggerConf.succeed("dig @127.0.0.1 printer.lan")
```
Comment (aanderse) [inline nixos/tests/blocky.nix:52]: Maybe something along these lines...
```
dig @127.0.0.1 +short -x 192.168.178.3 | grep -qF printer.lan
```

File: nixos/tests/blocky.nix [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    biggerConf = { pkgs, ... }: {
+      environment.systemPackages = [ pkgs.dnsutils ];
+      services.blocky = {
+        enable = true;
+
+        settings = {
+          customDNS = {
+            mapping = {
+              "printer.lan" = "192.168.178.3,2001:0db8:85a3:08d3:1319:8a2e:0370:7344";
```
Comment (arthurhenrique) [inline nixos/tests/blocky.nix:28]: is this a hard coded address? 👀

Comment (ratsclub) [reply nixos/tests/blocky.nix]: It is based on the [this example](https://github.com/0xERR0R/blocky/blob/64c198529136b07dfa9146c4a545e9481f2e06e7/docs/config.yml#L18-L23).

Comment (aanderse) [reply nixos/tests/blocky.nix]: Yeah, just to test that arbitrary custom DNS entries work.

**Comment by aanderse** [comment]
Looks great @ratsclub! Please add the test to all-tests.nix and in the package (`passthru.tests`) and we're good to merge.

You built the manual to ensure everything was fine with the release notes, right?

**Comment by ratsclub** [comment]
@aanderse I rebased the docs and rebuilt them.

File: nixos/tests/blocky.nix (hunk 4 above)
Comment (aanderse) [reply nixos/tests/blocky.nix:52]: This doesn't actually prove that address mapping works. You should put in what I had.

File: nixos/tests/blocky.nix (hunk 3 above)
Comment (aanderse) [reply nixos/tests/blocky.nix:5]: Again, I think you should drop this additional machine. It is entirely redundant to `biggerConf`.

**Review by aanderse — CHANGES_REQUESTED** [review]

**Comment by ratsclub** [comment]
Sorry, I messed up the rebase :( I'll fix it

**Comment by ratsclub** [comment]
@aanderse do you think I should improve something else?

**Review by aanderse — APPROVED** [review]
Perfect!

Sorry for the delay.

**Pull request merged**